### PR TITLE
#168781687 - Send Notifications

### PR DESCRIPTION
**Pull request by nignanthomas**
#### What does this PR do?
Send Notifications on request creation/edition/approval/rejection.

#### Description of Task to be completed?
- Create notifications table
- Add emailNotif column to users
- Create helpers to save notifications and send email notifications
- Create routes to getNotifications and mark as read/unread
- Create controllers/routes to switch on/off email notifications

#### How should this be manually tested?
- Clone the repository from [here](https://github.com/andela/caret-bn-backend.git)
- Browse to the repo directory
- navigate to ft-send-notification-168781687 branch
- Run `npm install`
- run `sequelize db:migrate`
- run `sequelize db:seed:all`
- run `npm run dev`

- To approve a request
Log in as a manager
Send a PATCH request to http://localhost:port/api/v1/requests/approve/requestId
The requester will receive an email notification.
- To reject a request
Log in as a manager
Send a PATCH request to http://localhost:port/api/v1/requests/reject/requestId
The requester will receive an email notification.

- To switch on/off the email notifications
Send a PATCH request to http://localhost:port/api/v1/users/email-notification

- To mark the notifications as read/unread
Send a PATCH request to http://localhost:port/api/v1/notifications/notificationId/mark

- To view notifications
Send a GET request to http://localhost:port/api/v1/notifications
Send a GET request to http://localhost:port/api/v1/notifications/notificationId

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[#168781687](https://www.pivotaltracker.com/story/show/168781687)

#### Screenshots (if appropriate)
Screen Shot 2019-10-29 at 12 32 57
Screen Shot 2019-10-29 at 12 34 40
Screen Shot 2019-10-29 at 12 33 10
Screen Shot 2019-10-29 at 12 29 20
Screen Shot 2019-10-29 at 12 31 23
Screen Shot 2019-10-29 at 12 31 11
Screen Shot 2019-10-29 at 12 30 00
Screen Shot 2019-10-29 at 12 29 46

#### Questions:
N/A

### CODE REVIEW

**Review by alainmateso — CHANGES_REQUESTED** [review]
Awesome job @nignanthomas . I have one small change to propose though. Kindly change the email subject when a request has been created. It seems like it sends an email with a subject for rejected whereas the email content is for created.
Screen Shot 2019-10-30 at 10 11 56

**Review by Cheza-Dzabala — APPROVED** [review]

**Review by alainmateso — APPROVED** [review]

File: src/controllers/destinationController.js [hunk 1]
```diff
@@ -1,12 +1,18 @@
...
 import requestServices from '../services/requestServices/requestServices';
 import Utilities from '../utils/index';
+import notifServices from '../services/notifServices';
+import notifSender from '../helpers/notifSender';
+
+const { notifSaver, notifBuilder } = notifServices;
 
 export default class destinationController {
-  static async storeDestination(res, body, user, request) {
+  static async storeDestination(req, res, body, user, request) {
```
Comment (nakiwuge) [inline src/controllers/destinationController.js]: remove this line

File: src/controllers/destinationController.js [hunk 2]
```diff
@@ -1,12 +1,18 @@
...
 
     const { destinations } = body;
```
Comment (nakiwuge) [inline src/controllers/destinationController.js]: remove this line

File: src/controllers/notificationController.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
+import strings from '../utils/stringsUtil';
+import responseUtil from '../utils/responseUtil';
+import notifServices from '../services/notifServices';
+
+const { allNotif, oneNotif, updateNotif } = notifServices;
+const { NOTIF_NOT_FOUND, NOTIF_FOUND } = strings.notifications;
+
```
Comment (nakiwuge) [inline src/controllers/notificationController.js:7]: remove this line

File: src/helpers/notifSender.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
+import nodemailer from 'nodemailer';
+import dotenv from 'dotenv';
+import models from '../database/models';
+
+dotenv.config();
+
+const transporter = nodemailer.createTransport({
+  service: 'gmail',
+  auth: {
+    user: 'caretdevs@gmail.com',
```
Comment (nakiwuge) [inline src/helpers/notifSender.js]: add the email to the .env file

File: src/helpers/notifSender.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+const notifSender = async (subject, request, userId, notification, APP_URL_BACKEND) => {
+  const user = await models.users.findOne({
+    where: { id: userId }
+  });
+
+  if (user.emailNotif) {
+    const { email, username } = user;
+    const msg = {
+      to: email,
+      from: 'caretdevs@gmail.com',
```
Comment (nakiwuge) [inline src/helpers/notifSender.js]: add the email to the .env file

File: src/helpers/notifSender.js [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+      subject,
+      html: `<div style="font-family:Avenir,Helvetica,sans-serif;box-sizing:border-box;padding:35px;">
+        <h1 style="color: #848484;">Barefoot Nomad</h1>
+        <p style="font-family:Avenir,Helvetica,sans-serif;box-sizing:border-box;color:#74787e;font-size:16px;line-height:1.5em;margin-top:0;text-align:left">
+          Welcome ${username},
+          <br> We are happy to be with you. This is to notify you that a request has been ${notification.activity} .
+          <br> Click the button below to view the request.
```
Comment (nakiwuge) [inline src/helpers/notifSender.js]: Remove the welcome since the user is not new to the application

File: src/helpers/notifSender.js [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    user: 'caretdevs@gmail.com',
+    pass: process.env.EMAIL_PASSWORD
+  }
+});
+
+const notifSender = async (subject, request, userId, notification, APP_URL_BACKEND) => {
+  const user = await models.users.findOne({
+    where: { id: userId }
+  });
+
```
Comment (nakiwuge) [inline src/helpers/notifSender.js]: create a helper function that can be reused to find a user. Something like this
 ```
const getUser=(dataObj)=>{
const user = await models.users.findOne({
    where: dataObj
  });
if (!user){return error}
return user
}
```

File: src/middlewares/isApproved.js [hunk 8]
```diff
@@ -0,0 +1,20 @@
+import strings from '../utils/stringsUtil';
+import responseError from '../utils/responseError';
+import models from '../database/models';
+
+const isApproved = (req, res, next) => {
+  models.requests.findOne({ where: { id: req.params.id } })
```
Comment (nakiwuge) [inline src/middlewares/isApproved.js]: Create a helper function for this method. You can use something like 
```
const getRequest=(dataObj)=>{
 const request = models.requests.findOne({
 where: dataObj
if(!request){
return error
}
return request
}
```

**Review by nakiwuge — CHANGES_REQUESTED** [review]
@nignanthomas, great work. I have left a few comments kindly check them out.

**Review by nakiwuge — APPROVED** [review]

**Pull request merged**